### PR TITLE
[NTOS:IO] Implement IoConnectInterruptEx for fully specific interrupt types

### PR DESCRIPTION
**Pull request by DarkFire01**
## Purpose

This gives ReactOS the ability to load various modern drivers that use IoConnectInterruptEx

various drivers work after this such as the serial.sys sample driver from MS and many more KMDF drivers from later windows versions

## Proposed changes

- IoConnectInterruptEx for CONNECT_FULLY_SPECIFIED

### CODE REVIEW

**Comment by JoachimHenze** [comment]
Please provide a specific driver that you used for your testing and that would work now!

**Comment by binarymaster** [comment]
> various drivers work after this such as the serial.sys sample driver from MS and many more KMDF drivers from later windows versions

You may want to create a Jira ticket to track the progress for a selected driver.

File: ntoskrnl/io/iomgr/irq.c [hunk 1]
```diff
@@ -166,4 +166,67 @@ IoDisconnectInterrupt(PKINTERRUPT InterruptObject)
     ExFreePool(IoInterrupt); // ExFreePoolWithTag(IoInterrupt, TAG_KINTERRUPT);
 }
 
+#if (NTDDI_VERSION >= NTDDI_LONGHORN) || defined(__REACTOS__)
```
Comment (Extravert-ir) [inline ntoskrnl/io/iomgr/irq.c:169]: I don't think this needs to be guarded with ifdefs

File: ntoskrnl/io/iomgr/irq.c [hunk 2]
```diff
@@ -166,4 +166,67 @@ IoDisconnectInterrupt(PKINTERRUPT InterruptObject)
...
+                                LocParameters.FullySpecified.ShareVector,
+                                LocParameters.FullySpecified.ProcessorEnableMask,
+                                LocParameters.FullySpecified.FloatingSave);
+    DPRINT("IopConnectInterruptEx_FullySpecific: has failed with status %X", Status);
+    return Status;
+}
+
+NTSTATUS
+NTAPI
+IoConnectInterruptEx(_Inout_ PIO_CONNECT_INTERRUPT_PARAMETERS Parameters)
```
Comment (Extravert-ir) [inline ntoskrnl/io/iomgr/irq.c:198]:
```suggestion
IoConnectInterruptEx(
    _Inout_ PIO_CONNECT_INTERRUPT_PARAMETERS Parameters)
```

File: ntoskrnl/io/iomgr/irq.c [hunk 3]
```diff
@@ -166,4 +166,67 @@ IoDisconnectInterrupt(PKINTERRUPT InterruptObject)
     ExFreePool(IoInterrupt); // ExFreePoolWithTag(IoInterrupt, TAG_KINTERRUPT);
 }
 
+#if (NTDDI_VERSION >= NTDDI_LONGHORN) || defined(__REACTOS__)
+NTSTATUS
+NTAPI
```
Comment (Extravert-ir) [inline ntoskrnl/io/iomgr/irq.c:171]:
```suggestion
```

File: ntoskrnl/io/iomgr/irq.c [hunk 4]
```diff
@@ -166,4 +166,67 @@ IoDisconnectInterrupt(PKINTERRUPT InterruptObject)
...
+IopConnectInterruptEx_FullySpecific(_Inout_ PIO_CONNECT_INTERRUPT_PARAMETERS Parameters)
```
Comment (Extravert-ir) [inline ntoskrnl/io/iomgr/irq.c:172]: Underscore is not usually used in names. I suggest `IopConnectInterruptFullySpecific`

**Review by Extravert-ir — CHANGES_REQUESTED** [review]

File: ntoskrnl/io/iomgr/irq.c (hunk 1 above)
Comment (DarkFire01) [reply ntoskrnl/io/iomgr/irq.c:169]: Just the private function? Or the whole thing?

File: ntoskrnl/io/iomgr/irq.c (hunk 2 above)
Comment (HBelusca) [reply ntoskrnl/io/iomgr/irq.c:198]: (NOTE: For this function and the two others introduced there.)

File: ntoskrnl/io/iomgr/irq.c (hunk 1 above)
Comment (Extravert-ir) [reply ntoskrnl/io/iomgr/irq.c]: The whole guard is not needed

File: ntoskrnl/io/iomgr/irq.c [hunk 5]
```diff
@@ -166,4 +166,67 @@ IoDisconnectInterrupt(PKINTERRUPT InterruptObject)
...
 
+NTSTATUS
+IopConnectInterruptExFullySpecific(
+    _Inout_ PIO_CONNECT_INTERRUPT_PARAMETERS Parameters)
+{
+    NTSTATUS Status;
+    IO_CONNECT_INTERRUPT_PARAMETERS LocParameters;
+    PAGED_CODE();
+
+    LocParameters = *Parameters;
```
Comment (Extravert-ir) [inline ntoskrnl/io/iomgr/irq.c:177]: Missed that at first. Why do you copy the structure here?

File: ntoskrnl/io/iomgr/irq.c [hunk 6]
```diff
@@ -166,4 +166,64 @@ IoDisconnectInterrupt(PKINTERRUPT InterruptObject)
...
 }
 
+NTSTATUS
+IopConnectInterruptExFullySpecific(
+    _Inout_ PIO_CONNECT_INTERRUPT_PARAMETERS Parameters)
+{
+    NTSTATUS Status;
+    PAGED_CODE();
+
+    /* Only fully FullySpecified interrupts this can call the legacy routine */
```
Comment (binarymaster) [inline ntoskrnl/io/iomgr/irq.c:176]: Please improve the wording of this comment somehow, it's not clear for me.

File: ntoskrnl/io/iomgr/irq.c [hunk 7]
```diff
@@ -166,4 +166,64 @@ IoDisconnectInterrupt(PKINTERRUPT InterruptObject)
...
+    PAGED_CODE();
+    switch (Parameters->Version)
+    {
+        case CONNECT_FULLY_SPECIFIED:
+            return IopConnectInterruptExFullySpecific(Parameters);
+            break;
+        case CONNECT_FULLY_SPECIFIED_GROUP:
+            //TODO: We don't do anything for the group type
+            return IopConnectInterruptExFullySpecific(Parameters);
+            break;
```
Comment (binarymaster) [inline ntoskrnl/io/iomgr/irq.c:206]:
```suggestion
        case CONNECT_FULLY_SPECIFIED:
            return IopConnectInterruptExFullySpecific(Parameters);
        case CONNECT_FULLY_SPECIFIED_GROUP:
            //TODO: We don't do anything for the group type
            return IopConnectInterruptExFullySpecific(Parameters);
```

**Review by Extravert-ir — APPROVED** [review]

**Review by binarymaster — APPROVED** [review]

**Pull request merged**